cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/NamespaceSourceControlMetadataRefresher.java

```diff
@@ -0,0 +1,210 @@
+    return NamespaceSourceControlMetadataStore.create(context);
+  }
+
+  private RepositoryTable getRepositoryTable(StructuredTableContext context)
+      throws TableNotFoundException {
+    return new RepositoryTable(context);
+  }
+
+
+  // TODO(CDAP-21020): Refactor this function
+  private RepositoryMeta getRepositoryMeta(NamespaceId namespace)
+      throws RepositoryNotFoundException {
+    return TransactionRunners.run(transactionRunner, context -> {
+      RepositoryTable table = getRepositoryTable(context);
+      RepositoryMeta repoMeta = table.get(namespace);
+      if (repoMeta == null) {
+        throw new RepositoryNotFoundException(namespace);
+      }
+
+      return repoMeta;
+    }, RepositoryNotFoundException.class);
+  }
+
+  public long refreshMetadata() throws NotFoundException {
+    LOG.info(
+        "Started NamespaceSourceControlMetadataRefreshService manual refresh for namespace {}",
+        namespaceId);
+    long refreshStartTime = System.currentTimeMillis();
+
+    // Getting repository config for a specific namespace
+    RepositoryConfig repoConfig = getRepositoryMeta(namespaceId).getConfig();
+    // Listing the apps from remote repo
+    RepositoryAppsResponse repositoryAppsResponse = sourceControlOperationRunner.list(
+        new NamespaceRepository(namespaceId, repoConfig));
+
+    Set<String> repoFileNames = repositoryAppsResponse.getApps().stream()
+        .map(RepositoryApp::getName)
+        .collect(Collectors.toCollection(HashSet::new));
+
+    LOG.info(
+        "Done listing apps from repo in NamespaceSourceControlMetadataRefreshService manual refresh for namespace {} with repo size {}",
+        namespaceId, repoFileNames.size());
+
+    TransactionRunners.run(transactionRunner, context -> {
+      RepositorySourceControlMetadataStore repoMetadataStore = getRepoSourceControlMetadataStore(
+          context);
+
+      // Cleaning up the repo source control metadata table
+      cleanupRepoSourceControlMetaInBatchesWithNoTransaction(namespaceId.getNamespace(),
+          repoFileNames, repoMetadataStore);
+
+      LOG.info(
+          "Done cleaning up apps from repo in NamespaceSourceControlMetadataRefreshService manual refresh for namespace {}",
+          namespaceId);
+
+      // Updating the namespace and repo source control metadata table
+      NamespaceSourceControlMetadataStore namespaceMetadataStore =
+          getNamespaceSourceControlMetadataStore(context);
+      for (RepositoryApp repoApp : repositoryAppsResponse.getApps()) {
+        updateSourceControlMetaWithNoTransaction(
+            new ApplicationReference(namespaceId.getNamespace(), repoApp.getName()),
+            repoApp.getFileHash(),
+            refreshStartTime, repoMetadataStore, namespaceMetadataStore);
+      }
+
+      LOG.info(
+          "Finished NamespaceSourceControlMetadataRefreshService manual refresh for namespace {}",
+          namespaceId);
+    });
+    lastRefreshTime.set(refreshStartTime);
```

Comment:
if you are using a atomic at the parent, no need for it to be set here

Reply:
didn't get this. I am created a map in sourcecontrolmetadatarefresher which maps namespace to last refresh time. I am not setting anything there

Reply:
ok makes sense